File: leetcode/medium/513-findBottomLeftTreeValue.js
// Given a binary tree, find the leftmost value in the last row of the tree.

// Example 1:
// Input:

//     2
//    / \
//   1   3

// Output:
// 1
// Example 2:
// Input:

//         1
//        / \
//       2   3
//      /   / \
//     4   5   6
//        /
//       7

// Output:
// 7

// first pass solution using a queue and BFS to find left most node on each level of tree
var findBottomLeftValue = function(root) {
    let levels = [];
    let queue = [root];
    
    while (queue.length > 0) {
        levels.push(queue[0].val);
        let len = queue.length;
        
        for (let i = 0; i < len; i++) {
            let curr = queue.shift();
            if (curr.left) queue.push(curr.left);
            if (curr.right) queue.push(curr.right);
        }
        
    }
    return levels[levels.length - 1];
};

File: leetcode/medium/513-findBottomLeftTreeValue.java
// first pass solution using BFT
class Solution {
    public int findBottomLeftValue(TreeNode root) {
        LinkedList<TreeNode> queue = new LinkedList<>();
        int output = -1;
        
        if (root != null) queue.addFirst(root);
        
        while (!queue.isEmpty()) {
            int len = queue.size();
            output = queue.peekFirst().val;
            
            for (int i = 0; i < len; i++) {
                TreeNode currNode = queue.removeFirst();
                
                if (currNode.left != null) queue.addLast(currNode.left);
                if (currNode.right != null) queue.addLast(currNode.right);
            }
        }
        return output;
    }
}

// second pass using DFT
class Solution {
    int maxLevel = -1;
    int output = -1;
    
    public int findBottomLeftValue(TreeNode root) {
        if (root == null) return -1;
        
        DFT(root, 0);
        
        return output;
    }
    
    public void DFT(TreeNode root, int level) {
        if (level > maxLevel) {
            output = root.val;
            maxLevel = level;
        }
        
        if (root.left != null) DFT(root.left, level + 1);
        if (root.right != null) DFT(root.right, level + 1);
    }
}
